Snowflake.Data/Client/SnowflakeCredentialManagerFactory.cs

```diff
@@ -0,0 +1,95 @@
+{
+    public class SnowflakeCredentialManagerFactory
+    {
+        private static readonly SFLogger s_logger = SFLoggerFactory.GetLogger<SnowflakeCredentialManagerFactory>();
+
+        private static readonly object s_credentialManagerLock = new object();
+
+        private static ISnowflakeCredentialManager s_credentialManager;
+        private static bool s_isDefaultCredentialManager = false;
+
+        internal static string BuildCredentialKey(string host, string user, TokenType tokenType, string authenticator = null)
+        {
+            return $"{host.ToUpper()}:{user.ToUpper()}:{SFEnvironment.DriverName}:{tokenType.ToString().ToUpper()}:{authenticator?.ToUpper() ?? string.Empty}";
+        }
+
+        public static void UseDefaultCredentialManager()
+        {
+            SetCredentialManager(GetDefaultCredentialManager());
+        }
+
+        public static void UseInMemoryCredentialManager()
+        {
+            SetCredentialManager(SFCredentialManagerInMemoryImpl.Instance);
+        }
+
+        public static void UseFileCredentialManager()
+        {
+            SetCredentialManager(SFCredentialManagerFileImpl.Instance);
+        }
+
+        public static void UseWindowsCredentialManager()
+        {
+            SetCredentialManager(SFCredentialManagerWindowsNativeImpl.Instance);
+        }
+
+        public static void SetCredentialManager(ISnowflakeCredentialManager customCredentialManager)
+        {
+            lock (s_credentialManagerLock)
+            {
+                if (customCredentialManager == null)
+                {
+                    throw new SnowflakeDbException(SFError.INTERNAL_ERROR,
+                        "Credential manager cannot be null. If you want to use the default credential manager, please call the UseDefaultCredentialManager method.");
+                }
+
+                if (customCredentialManager == s_credentialManager)
+                {
+                    s_logger.Info($"Credential manager is already set to: {customCredentialManager.GetType().Name}");
+                    return;
+                }
+
+                s_isDefaultCredentialManager = customCredentialManager == GetDefaultCredentialManager();
+                s_logger.Info($"Setting the credential manager: {customCredentialManager.GetType().Name}");
+                s_credentialManager = customCredentialManager;
+            }
+        }
+
+        public static ISnowflakeCredentialManager GetCredentialManager()
+        {
+            if (s_credentialManager == null)
+            {
+                lock (s_credentialManagerLock)
+                {
+                    if (s_credentialManager == null)
+                    {
+                        s_isDefaultCredentialManager = true;
+                        s_credentialManager = GetDefaultCredentialManager();
+                    }
+                }
+            }
+            var typeCredentialText = s_isDefaultCredentialManager ? "default" : "custom";
```

Comment:
I'm thinking of that, and maybe you were right to have to have default credentialManager on a static filed in the class. Then we could remove s_isDefaultCredentialManager bool flag and easily check if the objects are the same.
We could add a field `private static readonly ISnowflakeCredentialManager s_defaultCredentialManager = GetDefaultCredentialManager();`
Then here we could check here if the implementation is default by comparing to this object.
I would only first assign the s_credentialManager to a variable, check on this variable and return its value, to avoid the possibility, that the manager was replaced when we were checking if it is default.

Reply:
Done, thank you!!